### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry): the big Zariski site

### PR DESCRIPTION
**Pull request by joelriou**
This PR defines the Zariski topology on the category of schemes. It may have some applications, but it should also be considered as a test case for the Grothendieck topologies API in order to check that is general enough, in particular in terms of universes: a priori, the category which shall form the small étale site of a scheme will have the same universe parameters.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/Sites/BigZariski.lean [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  coverings Y S := ∃ (U : OpenCover.{u} Y), S = Presieve.ofArrows U.obj U.map
+  has_isos Y X f _ := ⟨openCoverOfIsIso f, (Presieve.ofArrows_pUnit _).symm⟩
+  pullbacks := by
+    rintro Y X f _ ⟨U, rfl⟩
+    exact ⟨U.pullbackCover' f, (Presieve.ofArrows_pullback _ _ _).symm⟩
+  Transitive := by
+    rintro X _ T ⟨U, rfl⟩ H
+    let V : ∀ (j : U.J), OpenCover (U.obj j) := fun j => (H (U.map j) ⟨j⟩).choose
+    exact ⟨U.bind V, Presieve.ofArrows_bind' U.obj U.map (fun i => (V i).obj)
+      (fun i => (V i).map) T (fun j => (H (U.map j) ⟨j⟩).choose_spec)⟩
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Sites/BigZariski.lean:37]:
```suggestion
    rintro X _ T ⟨U, rfl⟩ H
    choose V hV using H
    use U.bind (fun j => V (U.map j) ⟨j⟩)
    simpa only [OpenCover.bind, ← hV] using Presieve.ofArrows_bind U.obj U.map _
      (fun _ f H => (V f H).obj) (fun _ f H => (V f H).map)
```
I don't think we need `Presieve.ofArrows_bind'`?

File: Mathlib/AlgebraicGeometry/Sites/BigZariski.lean [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Sites/BigZariski.lean:38]: We should have this result
```
example {Y : Scheme.{u}} (U : OpenCover.{v} Y) :
  zariskiPretopology Y (Presieve.ofArrows U.obj U.map)
```
But I'm not sure if it is easy enough to fit in this relatively small PR.
Maybe add the `OpenCover.{u} Y` version and add a TODO note?

File: Mathlib/AlgebraicGeometry/Sites/BigZariski.lean (hunk 1 above)
Comment (joelriou) [reply Mathlib/AlgebraicGeometry/Sites/BigZariski.lean:37]: I was just not good enough to use `ofArrows_bind` :-(

File: Mathlib/AlgebraicGeometry/Sites/BigZariski.lean (hunk 2 above)
Comment (joelriou) [reply Mathlib/AlgebraicGeometry/Sites/BigZariski.lean:50]: Thanks for the suggestion. I have included the `OpenCover.{u}` version as well as the following:
```lean
lemma zariskiTopology_openCover {Y : Scheme.{u}} (U : OpenCover.{v} Y) :
    zariskiTopology Y (Sieve.generate (Presieve.ofArrows U.obj U.map))
```

**Comment by erdOne** [comment]
LGTM
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**